url89cafed56a_AlexDiru_esper_compiler_final_tgz-pJ8-compiler_ProgramJ8/src/compiler/EsperParser.java

```diff
@@ -0,0 +1,1997 @@
+    };
+
+
+    // $ANTLR start "expr"
+    // C:\\Users\\Alex\\Dropbox\\EclipseWorkspace\\esper-compiler\\src\\compiler\\Esper.g:31:1: expr : ( ( PLUS | MINUS ) ^ expr expr | term | STRING );
+    public final EsperParser.expr_return expr() throws RecognitionException {
+        EsperParser.expr_return retval = new EsperParser.expr_return();
+        retval.start = input.LT(1);
+
+
+        Object root_0 = null;
+
+        Token set11=null;
+        Token STRING15=null;
+        EsperParser.expr_return expr12 =null;
+
+        EsperParser.expr_return expr13 =null;
+
+        EsperParser.term_return term14 =null;
+
+
+        Object set11_tree=null;
+        Object STRING15_tree=null;
+
+        try {
+            // C:\\Users\\Alex\\Dropbox\\EclipseWorkspace\\esper-compiler\\src\\compiler\\Esper.g:31:9: ( ( PLUS | MINUS ) ^ expr expr | term | STRING )
+            int alt3=3;
+            switch ( input.LA(1) ) {
+            case MINUS:
+            case PLUS:
+                {
+                alt3=1;
+                }
+                break;
+            case DIGITS:
+            case DIV:
+            case IDENTIFIER:
+            case INFINITY:
+            case MULT:
+            case NEGATIVEINFINITY:
+            case NULLITY:
+                {
+                alt3=2;
+                }
+                break;
+            case STRING:
+                {
+                alt3=3;
+                }
+                break;
+            default:
+                NoViableAltException nvae =
+                    new NoViableAltException("", 3, 0, input);
+
+                throw nvae;
+
+            }
+
+            switch (alt3) {
+                case 1 :
+                    // C:\\Users\\Alex\\Dropbox\\EclipseWorkspace\\esper-compiler\\src\\compiler\\Esper.g:31:11: ( PLUS | MINUS ) ^ expr expr
+                    {
+                    root_0 = (Object)adaptor.nil();
+
+
+                    set11=(Token)input.LT(1);
+
+                    set11=(Token)input.LT(1);
+
+                    if ( input.LA(1)==MINUS||input.LA(1)==PLUS ) {
+                        input.consume();
+                        root_0 = (Object)adaptor.becomeRoot(
+                        (Object)adaptor.create(set11)
+                        , root_0);
+                        state.errorRecovery=false;
+                    }
+                    else {
+                        MismatchedSetException mse = new MismatchedSetException(null,input);
+                        throw mse;
+                    }
+
+
+                    pushFollow(FOLLOW_expr_in_expr200);
+                    expr12=expr();
+
+                    state._fsp--;
+
+                    adaptor.addChild(root_0, expr12.getTree());
+
+                    pushFollow(FOLLOW_expr_in_expr202);
+                    expr13=expr();
+
+                    state._fsp--;
+
+                    adaptor.addChild(root_0, expr13.getTree());
+
+                    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 533:568, 670:705.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.